### DISCUSSION TITLE
单机版和分布式版查询速度相差10倍？

### DISCUSSION THREAD

**Discussion by zyy20191** [milvus-io/milvus#17961] [Q&A and General discussion]
同样的操作，相同的数据，我用单机版查询速度是200ms，分布式版是20ms，请问是什么原因呢？无论是否建立索引，速度差距都是这么多

**Comment by yhmo** [comment]
按理说不应该，请详细说明一下你的操作：
1. milvus版本号
2. 数据行数
3. 客户端脚本

**Reply by zyy20191** [reply]
1、版本号：milvusdb/milvus:v2.0.0-rc8-20211104-d1f4106
2、数据行数：200
3、
# 创建entity
book_id = FieldSchema(
    name="nlp_id",
    dtype=DataType.INT64,
    is_primary=True,
	auto_id=False
    )
book_intro = FieldSchema(
    name="nlp_intro",
    dtype=DataType.FLOAT_VECTOR,
    dim=8
    )

# 创建collection
collection_name = "nlp5"
schema = CollectionSchema(
    fields=[book_id, book_intro],
    description="nlp"
    )
collection = Collection(
    name=collection_name,
    schema=schema
    )

# 插入数据
booK = Collection(collection_name)
ls = []
with open("insert2.csv", "r", encoding="utf-8") as f:
    a = f.readline()
    while a:
        a = a.strip()
        ls.append(a)
        a = f.readline()

entities = [
    [i for i in range(200)],  # field pk
    [[random.random() for _ in range(8)] for _ in range(200)],  # field embeddings
]

isIT = booK.insert(entities)
print("插入数据：", isIT)

###########################################
# 查询数据
search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
startTime = datetime.datetime.now()
rul = booK.search(
	data=[[1.1233, 2.132, 2.1328, 2.1326, 2.1324, 2.1326, 2.1324, 2.13212]],
	anns_field="nlp_intr",
	param=search_params,
	limit=3
)
endTime = datetime.datetime.now()
print("cost {}s time.".format((endTime - startTime).total_seconds()))

###########################################
用此方法打印出的时间均在200ms左右

**Comment by yhmo** [comment]
我用下面的脚本测试：
```
import random
import datetime

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
    Index
)

HOST = '127.0.0.1'
PORT = '19530'

if __name__ == '__main__':
    connections.connect(host=HOST, port=PORT)

    book_id = FieldSchema(
        name="nlp_id",
        dtype=DataType.INT64,
        is_primary=True,
        auto_id=False
    )
    book_intro = FieldSchema(
        name="nlp_intro",
        dtype=DataType.FLOAT_VECTOR,
        dim=8
    )

    collection_name = "nlp5"
    schema = CollectionSchema(
        fields=[book_id, book_intro],
        description="nlp"
    )
    collection = Collection(
        name=collection_name,
        schema=schema
    )

    booK = Collection(collection_name)

    entities = [
        [i for i in range(200)],  # field pk
        [[random.random() for _ in range(8)] for _ in range(200)],  # field embeddings
    ]

    isIT = booK.insert(entities)
    print("插入数据：", isIT)

    booK.load()

    for i in range(10):
        search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
        startTime = datetime.datetime.now()
        rul = booK.search(
            data=[[1.1233, 2.132, 2.1328, 2.1326, 2.1324, 2.1326, 2.1324, 2.13212]],
            anns_field="nlp_intro",
            param=search_params,
            limit=3
        )
        endTime = datetime.datetime.now()
        print("cost {}ms time.".format((endTime - startTime).total_seconds()*1000))
```

**Reply by yhmo** [reply]
首先使用2.0.1的单机模式：
```
wget https://github.com/milvus-io/milvus/releases/download/v2.0.1/milvus-standalone-docker-compose.yml -O docker-compose.yml
docker-compose up -d
```
跑这个脚本得出：
```
插入数据： (insert count: 200, delete count: 0, upsert count: 0, timestamp: 434282522688880649)
cost 20.394ms time.
cost 18.601ms time.
cost 17.701ms time.
cost 17.507ms time.
cost 17.191000000000003ms time.
cost 14.9ms time.
cost 8.893ms time.
cost 6.707ms time.
cost 5.948ms time.
cost 5.155ms time.
```

然后使用2.0.1的cluster模式(在一台机器上起多个组件)：
```
wget https://github.com/milvus-io/milvus/releases/download/v2.0.1/milvus-cluster-docker-compose.yml -O docker-compose.yml
docker-compose up -d
```
然后跑同样的脚本：
```
插入数据： (insert count: 200, delete count: 0, upsert count: 0, timestamp: 434282632021803013)
cost 29.634ms time.
cost 20.028000000000002ms time.
cost 9.303ms time.
cost 10.411999999999999ms time.
cost 10.567ms time.
cost 10.606000000000002ms time.
cost 9.684999999999999ms time.
cost 10.616999999999999ms time.
cost 10.175ms time.
cost 10.321ms time.
```

都是插入200条数据后search 10次，两者在性能上并没有多少区别。可能是你测的时候被别的因素影响了，正常来讲这两者跑在同一台机器上的差别很小。

**Reply by zyy20191** [reply]
非常感谢，我将版本从v2.0.0升级到v2.0.2之后解决了问题。同样的操作，在2.0.2版本中速度可以达到20ms，所以我怀疑是安装时操作失误？亦或者版本BUG。
